File: web/app/services/campaignService.ts
/**
 * Campaign service — HTTP client for the /campaigns API routes.
 * @module services/campaignService
 */
import { ApiClient } from './api'
import type { CampaignFollowUp, CampaignVariantStats } from '~/types'

export type CampaignStatus = 'draft' | 'active' | 'completed' | 'paused' | 'cancelled'

export type QueueItemStatus = 'pending' | 'sending' | 'sent' | 'skipped' | 'failed'

export type CampaignProspect = {
  id: number
  name: string
  email?: string | null
  phone?: string | null
  city?: string | null
  category: string
  source: string
  confidence: number
  ab_variant?: string | null
}

export type CampaignResponse = {
  id: number
  user_id: number
  name: string
  description?: string | null
  status: CampaignStatus
  template_id?: number | null
  ab_template_id_b?: number | null
  send_delay_minutes: number
  follow_up_delay_days: number
  behavior_personalized_followups: boolean
  include_video: boolean
  max_emails_per_day?: number | null
  started_at?: string | null
  created_at: string
  updated_at?: string | null
  prospects_count: number
  /** Earliest still-pending send in the queue (naive UTC), or null when nothing is queued. */
  next_send_at?: string | null
}

export interface CampaignDetailResponse extends CampaignResponse {
  prospects: CampaignProspect[]
  follow_ups: CampaignFollowUp[]
}

export type CampaignListResponse = {
  campaigns: CampaignResponse[]
  total: number
  /** Count of the user's pending sends scheduled within the next 7 days. */
  upcoming_sends_7d?: number
}

export type CampaignStats = {
  campaign_id: number
  total_prospects: number
  total_emails_sent: number
  emails_delivered: number
  emails_opened: number
  emails_clicked: number
  emails_replied: number
  emails_bounced: number
  emails_failed: number
  delivery_rate: number
  open_rate: number
  click_rate: number
  reply_rate: number
  ab_stats?: CampaignVariantStats[] | null
}

export type CampaignCreatePayload = {
  name: string
  description?: string
  status?: CampaignStatus
  prospect_ids?: number[]
  template_id?: number
  ab_template_id_b?: number
  send_delay_minutes?: number
  max_emails_per_day?: number | null
}

export type CampaignUpdatePayload = {
  name?: string
  description?: string
  status?: CampaignStatus
}

export type CampaignSettingsPayload = {
  template_id?: number | null
  ab_template_id_b?: number | null
  disable_ab?: boolean
  send_delay_minutes?: number
  max_emails_per_day?: number | null
  clear_max_emails_per_day?: boolean
  behavior_personalized_followups?: boolean
  include_video?: boolean
  follow_ups?: Array<{ template_id: number; delay_days: number; position: number }>
}

export type CampaignLaunchPayload = {
  template_id?: number
  ab_template_id_b?: number
  follow_up_template_id?: number
  follow_up_delay_days?: number
  send_delay_minutes?: number
}

export type LaunchCampaignResponse = {
  success: boolean
  enqueued: number
  skipped_no_demo?: Array<{ id: number; name: string }>
  message: string
}

/** A single item in the campaign send queue. */
export type CampaignQueueItem = {
  id: number
  queue_type: 'initial' | 'followup'
  status: QueueItemStatus
  scheduled_at: string
  prospect_id: number
  prospect_name?: string | null
  prospect_email?: string | null
  ab_variant?: string | null
  follow_up_index: number
  email_log_id?: number | null
  /** Why a row was skipped (e.g. « Annulé manuellement », « Site démo expiré ») — null otherwise. */
  skip_reason?: string | null
}

export type CampaignQueueResponse = {
  pending_count: number
  items: CampaignQueueItem[]
}

/** Kind of link an email carries — only the website (demo) link exists today. */
export type ForecastLinkKind = 'website'

/** One scheduled send in the week-ahead forecast, across all campaigns. */
export type CampaignForecastItem = {
  queue_id: number
  scheduled_at: string
  campaign_id: number
  campaign_name: string
  prospect_id: number
  prospect_name?: string | null
  prospect_email?: string | null
  prospect_city?: string | null
  prospect_category: string
  queue_type: 'initial' | 'followup'
  follow_up_index: number
  ab_variant?: string | null
  /** 'pending' for an upcoming send; 'skipped' (with skip_reason) for one that will not go out. */
  status: QueueItemStatus
  skip_reason?: string | null
  /** The link the email will carry, its kind, and the reviewable site — null when no active demo. */
  link?: string | null
  link_kind?: ForecastLinkKind | null
  demo_site_id?: number | null
  site_reviewed_at?: string | null
}

export type CampaignForecastResponse = {
  items: CampaignForecastItem[]
  range_start: string
  range_end: string
}

export class CampaignService {
  /**
   * Create a new campaign.
   * @param data - Campaign creation payload.
   */
  static async create(data: CampaignCreatePayload): Promise<CampaignDetailResponse> {
    return ApiClient.post<CampaignDetailResponse>('/api/v1/campaigns', data)
  }

  /**
   * List campaigns with optional status filtering.
   * @param skip   - Pagination offset.
   * @param limit  - Max records to return.
   * @param status - Optional status filter.
   */
  static async list(skip: number = 0, limit: number = 100, status?: CampaignStatus): Promise<CampaignListResponse> {
    const params: URLSearchParams = new URLSearchParams({ skip: skip.toString(), limit: limit.toString() })
    if (status) params.append('status', status)
    return ApiClient.get<CampaignListResponse>(`/api/v1/campaigns?${params.toString()}`)
  }

  /**
   * Fetch a campaign by ID including prospects and follow-up sequence.
   * @param id - Campaign ID.
   */
  static async get(id: number): Promise<CampaignDetailResponse> {
    return ApiClient.get<CampaignDetailResponse>(`/api/v1/campaigns/${id}`)
  }

  /**
   * Update a campaign's name, description, or status.
   * @param id   - Campaign ID.
   * @param data - Fields to update.
   */
  static async update(id: number, data: CampaignUpdatePayload): Promise<CampaignDetailResponse> {
    return ApiClient.patch<CampaignDetailResponse>(`/api/v1/campaigns/${id}`, data)
  }

  /**
   * Update campaign send configuration (template, account, A/B, follow-ups).
   * Safe to call on active campaigns — changes apply to unsent items.
   * @param id       - Campaign ID.
   * @param settings - Configuration to update.
   */
  static async updateSettings(id: number, settings: CampaignSettingsPayload): Promise<CampaignDetailResponse> {
    return ApiClient.patch<CampaignDetailResponse>(`/api/v1/campaigns/${id}/settings`, settings)
  }

  /**
   * Permanently delete a campaign.
   * @param id - Campaign ID.
   */
  static async delete(id: number): Promise<void> {
    await ApiClient.delete(`/api/v1/campaigns/${id}`)
  }

  /**
   * Add prospects to a campaign.
   * @param campaignId  - Campaign ID.
   * @param prospectIds - IDs of prospects to add.
   */
  static async addProspects(campaignId: number, prospectIds: number[]): Promise<CampaignDetailResponse> {
    return ApiClient.post<CampaignDetailResponse>(`/api/v1/campaigns/${campaignId}/prospects`, {
      prospect_ids: prospectIds,
    })
  }

  /**
   * Remove a single prospect from a campaign.
   * @param campaignId - Campaign ID.
   * @param prospectId - Prospect ID to remove.
   */
  static async removeProspect(campaignId: number, prospectId: number): Promise<CampaignDetailResponse> {
    return ApiClient.delete<CampaignDetailResponse>(`/api/v1/campaigns/${campaignId}/prospects/${prospectId}`)
  }

  /**
   * Fetch aggregated statistics for a campaign (includes A/B breakdown when applicable).
   * @param campaignId - Campaign ID.
   */
  static async getStats(campaignId: number): Promise<CampaignStats> {
    return ApiClient.get<CampaignStats>(`/api/v1/campaigns/${campaignId}/stats`)
  }

  /**
   * Launch a campaign — populates the rate-limited send queue.
   * Uses the campaign's stored template/account when request fields are omitted.
   * @param campaignId - Campaign ID.
   * @param data       - Optional overrides for template, account, timing.
   */
  static async launch(campaignId: number, data: CampaignLaunchPayload = {}): Promise<LaunchCampaignResponse> {
    return ApiClient.post<LaunchCampaignResponse>(`/api/v1/campaigns/${campaignId}/launch`, data)
  }

  /**
   * Pause a running campaign — cancels all pending queue items.
   * @param campaignId - Campaign ID.
   */
  static async pause(campaignId: number): Promise<{ success: boolean; cancelled: number }> {
    return ApiClient.post<{ success: boolean; cancelled: number }>(`/api/v1/campaigns/${campaignId}/pause`, {})
  }

  /**
   * Resume a paused campaign — re-enqueues prospects not yet contacted.
   * @param campaignId - Campaign ID.
   */
  static async resume(campaignId: number): Promise<{ success: boolean; enqueued: number }> {
    return ApiClient.post<{ success: boolean; enqueued: number }>(`/api/v1/campaigns/${campaignId}/resume`, {})
  }

  /**
   * Immediately dispatch a follow-up email to a prospect (bypass delay).
   * Sends via the user's Resend configuration.
   * @param campaignId - Campaign ID.
   * @param prospectId - Prospect to target.
   * @param templateId - Template to use.
   */
  static async sendNow(
    campaignId: number,
    prospectId: number,
    templateId: number,
  ): Promise<{ success: boolean; email_log_id?: number; error?: string }> {
    return ApiClient.post(`/api/v1/campaigns/${campaignId}/send-now`, {
      prospect_id: prospectId,
      template_id: templateId,
    })
  }

  /**
   * Cancel a single pending queue item so it is not sent.
   * The item becomes « Ignoré » with a manual reason and can be re-sent later.
   * @param campaignId - Campaign ID.
   * @param queueId    - Queue item ID.
   */
  static async cancelQueueItem(
    campaignId: number,
    queueId: number,
  ): Promise<{ success: boolean; id: number; status: QueueItemStatus }> {
    return ApiClient.post(`/api/v1/campaigns/${campaignId}/queue/${queueId}/cancel`, {})
  }

  /**
   * Re-queue a skipped item so the worker sends it on its next tick (~1 min).
   * @param campaignId - Campaign ID.
   * @param queueId    - Queue item ID.
   */
  static async resendQueueItem(
    campaignId: number,
    queueId: number,
  ): Promise<{ success: boolean; id: number; status: QueueItemStatus; scheduled_at?: string }> {
    return ApiClient.post(`/api/v1/campaigns/${campaignId}/queue/${queueId}/resend`, {})
  }

  /**
   * Fetch the send queue for a campaign.
   * @param campaignId - Campaign ID.
   * @param params     - Optional filters and pagination.
   */
  static async getQueue(
    campaignId: number,
    params?: { status?: QueueItemStatus; limit?: number; offset?: number },
  ): Promise<CampaignQueueResponse> {
    const qs: URLSearchParams = new URLSearchParams()
    if (params?.status) qs.set('status', params.status)
    if (params?.limit !== undefined) qs.set('limit', String(params.limit))
    if (params?.offset !== undefined) qs.set('offset', String(params.offset))
    return ApiClient.get<CampaignQueueResponse>(`/api/v1/campaigns/${campaignId}/queue?${qs.toString()}`)
  }

  /**
   * Fetch the week-ahead send forecast across all of the user's campaigns.
   * @param startIso - Window start as an ISO datetime (the exact UTC instant of the local week start).
   * @param days     - Window width in days (default 7).
   */
  static async getForecast(startIso: string, days: number = 7): Promise<CampaignForecastResponse> {
    const qs: URLSearchParams = new URLSearchParams({ start: startIso, days: String(days) })
    return ApiClient.get<CampaignForecastResponse>(`/api/v1/campaigns/forecast?${qs.toString()}`)
  }
}

File: api/services/campaign_service.py
"""
Campaign service for managing email campaigns.
"""

from fastapi import HTTPException, status
from sqlalchemy.orm import Session, joinedload

from models.campaign import Campaign, CampaignStatus
from models.email_log import EmailLog
from models.prospect_db import ProspectDB
from schemas.campaign import (
    CampaignCreate,
    CampaignStats,
    CampaignUpdate,
    CampaignVariantStats,
)
from services.activity_log_service import CATEGORY_CAMPAIGN, STATUS_INFO, activity_log_service
from services.email_log_stats import aggregate_email_log_counts, compute_engagement_rates

# Campaign status → French verb for the activity feed.
_CAMPAIGN_STATUS_VERBS: dict[str, str] = {
    CampaignStatus.ACTIVE.value: "activée",
    CampaignStatus.PAUSED.value: "mise en pause",
    CampaignStatus.COMPLETED.value: "terminée",
    CampaignStatus.CANCELLED.value: "annulée",
    CampaignStatus.DRAFT.value: "repassée en brouillon",
}


class CampaignService:
    """Service for campaign management."""

    def create_campaign(self, db: Session, user_id: int, campaign_data: CampaignCreate) -> Campaign:
        """
        Create a new campaign.

        Args:
            db: Database session
            user_id: ID of the user creating the campaign
            campaign_data: Campaign creation data

        Returns:
            Created campaign
        """
        # Create campaign
        campaign = Campaign(
            user_id=user_id,
            name=campaign_data.name,
            description=campaign_data.description,
            status=campaign_data.status or CampaignStatus.DRAFT.value,
            template_id=campaign_data.template_id,
            ab_template_id_b=campaign_data.ab_template_id_b,
            send_delay_minutes=campaign_data.send_delay_minutes,
            max_emails_per_day=campaign_data.max_emails_per_day,
        )

        # Add prospects if provided
        if campaign_data.prospect_ids:
            prospects = (
                db.query(ProspectDB)
                .filter(ProspectDB.id.in_(campaign_data.prospect_ids), ProspectDB.user_id == user_id)
                .all()
            )
            campaign.prospects = prospects

        db.add(campaign)
        db.commit()
        db.refresh(campaign)

        activity_log_service.record(
            category=CATEGORY_CAMPAIGN,
            action="campaign_created",
            status=STATUS_INFO,
            title=f"Campagne créée · {campaign.name}",
            user_id=user_id,
            entity_type="campaign",
            entity_id=campaign.id,
        )

        return campaign

    def get_campaign(self, db: Session, campaign_id: int, user_id: int) -> Campaign | None:
        """
        Get a campaign by ID.

        Args:
            db: Database session
            campaign_id: Campaign ID
            user_id: ID of the user requesting the campaign

        Returns:
            Campaign if found and owned by user, None otherwise
        """
        campaign = (
            db.query(Campaign)
            .options(
                joinedload(Campaign.prospects),
                joinedload(Campaign.follow_ups),
            )
            .filter(Campaign.id == campaign_id, Campaign.user_id == user_id)
            .first()
        )

        return campaign

    def list_campaigns(
        self, db: Session, user_id: int, skip: int = 0, limit: int = 100, status: str | None = None
    ) -> tuple[list[Campaign], int]:
        """
        List campaigns for a user.

        Args:
            db: Database session
            user_id: ID of the user
            skip: Number of records to skip
            limit: Maximum number of records to return
            status: Filter by status (optional)

        Returns:
            Tuple of (campaigns list, total count)
        """
        query = db.query(Campaign).filter(Campaign.user_id == user_id)

        if status:
            query = query.filter(Campaign.status == status)

        total = query.count()
        campaigns = query.order_by(Campaign.created_at.desc()).offset(skip).limit(limit).all()

        return campaigns, total

    def update_campaign(
        self, db: Session, campaign_id: int, user_id: int, campaign_data: CampaignUpdate
    ) -> Campaign | None:
        """
        Update a campaign.

        Args:
            db: Database session
            campaign_id: Campaign ID
            user_id: ID of the user updating the campaign
            campaign_data: Campaign update data

        Returns:
            Updated campaign if found and owned by user, None otherwise
        """
        campaign = self.get_campaign(db, campaign_id, user_id)
        if not campaign:
            return None

        # Update fields
        previous_status = campaign.status
        update_data = campaign_data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(campaign, field, value)

        db.commit()
        db.refresh(campaign)

        if campaign.status != previous_status:
            verb = _CAMPAIGN_STATUS_VERBS.get(campaign.status, campaign.status)
            activity_log_service.record(
                category=CATEGORY_CAMPAIGN,
                action="campaign_status_changed",
                status=STATUS_INFO,
                title=f"Campagne {verb} · {campaign.name}",
                user_id=user_id,
                entity_type="campaign",
                entity_id=campaign.id,
            )

        return campaign

    def delete_campaign(self, db: Session, campaign_id: int, user_id: int) -> bool:
        """
        Delete a campaign.

        Args:
            db: Database session
            campaign_id: Campaign ID
            user_id: ID of the user deleting the campaign

        Returns:
            True if deleted, False if not found
        """
        campaign = self.get_campaign(db, campaign_id, user_id)
        if not campaign:
            return False

        campaign_name = campaign.name
        deleted_id = campaign.id
        db.delete(campaign)
        db.commit()

        activity_log_service.record(
            category=CATEGORY_CAMPAIGN,
            action="campaign_deleted",
            status=STATUS_INFO,
            title=f"Campagne supprimée · {campaign_name}",
            user_id=user_id,
            entity_type="campaign",
            entity_id=deleted_id,
        )

        return True

    def add_prospects_to_campaign(
        self, db: Session, campaign_id: int, user_id: int, prospect_ids: list[int]
    ) -> Campaign | None:
        """
        Add prospects to a campaign.

        Args:
            db: Database session
            campaign_id: Campaign ID
            user_id: ID of the user
            prospect_ids: List of prospect IDs to add

        Returns:
            Updated campaign if found, None otherwise

        Raises:
            HTTPException: If prospects not found or not owned by user
        """
        campaign = self.get_campaign(db, campaign_id, user_id)
        if not campaign:
            return None

        # Get prospects
        prospects = db.query(ProspectDB).filter(ProspectDB.id.in_(prospect_ids), ProspectDB.user_id == user_id).all()

        if len(prospects) != len(prospect_ids):
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, detail="Some prospects not found or not owned by user"
            )

        # Get existing prospect IDs
        existing_ids = {p.id for p in campaign.prospects}

        # Add only new prospects
        for prospect in prospects:
            if prospect.id not in existing_ids:
                campaign.prospects.append(prospect)

        db.commit()
        db.refresh(campaign)

        return campaign

    def remove_prospect_from_campaign(
        self, db: Session, campaign_id: int, user_id: int, prospect_id: int
    ) -> Campaign | None:
        """
        Remove a prospect from a campaign.

        Args:
            db: Database session
            campaign_id: Campaign ID
            user_id: ID of the user
            prospect_id: Prospect ID to remove

        Returns:
            Updated campaign if found, None otherwise
        """
        campaign = self.get_campaign(db, campaign_id, user_id)
        if not campaign:
            return None

        # Find and remove prospect
        for i, prospect in enumerate(campaign.prospects):
            if prospect.id == prospect_id:
                campaign.prospects.pop(i)
                break

        db.commit()
        db.refresh(campaign)

        return campaign

    def get_campaign_stats(self, db: Session, campaign_id: int, user_id: int) -> CampaignStats | None:
        """
        Get statistics for a campaign.

        Args:
            db: Database session
            campaign_id: Campaign ID
            user_id: ID of the user

        Returns:
            Campaign statistics if campaign found, None otherwise
        """
        campaign = self.get_campaign(db, campaign_id, user_id)
        if not campaign:
            return None

        # Count prospects
        total_prospects = len(campaign.prospects)

        counts = aggregate_email_log_counts(
            db,
            EmailLog.campaign_id == campaign_id,
            EmailLog.user_id == user_id,
        )
        rates = compute_engagement_rates(counts)

        # A/B breakdown (only when campaign has a B variant)
        ab_stats: list[CampaignVariantStats] | None = None
        if campaign.ab_template_id_b:
            ab_stats = []
            for variant in ("A", "B"):
                variant_counts = aggregate_email_log_counts(
                    db,
                    EmailLog.campaign_id == campaign_id,
                    EmailLog.user_id == user_id,
                    EmailLog.ab_variant == variant,
                )
                variant_rates = compute_engagement_rates(variant_counts)
                ab_stats.append(
                    CampaignVariantStats(
                        variant=variant,
                        sent=variant_counts.sent,
                        delivered=variant_counts.delivered,
                        opened=variant_counts.opened,
                        clicked=variant_counts.clicked,
                        replied=variant_counts.replied,
                        open_rate=variant_rates.open_rate,
                        click_rate=variant_rates.click_rate,
                        reply_rate=variant_rates.reply_rate,
                    )
                )

        return CampaignStats(
            campaign_id=campaign_id,
            total_prospects=total_prospects,
            total_emails_sent=counts.sent,
            emails_delivered=counts.delivered,
            emails_opened=counts.opened,
            emails_clicked=counts.clicked,
            emails_replied=counts.replied,
            emails_bounced=counts.bounced,
            emails_failed=counts.failed,
            delivery_rate=rates.delivery_rate,
            open_rate=rates.open_rate,
            click_rate=rates.click_rate,
            reply_rate=rates.reply_rate,
            ab_stats=ab_stats,
        )


# Singleton instance
campaign_service = CampaignService()
